File: src/java/br/com/lett/crawlernode/crawlers/corecontent/brasil/BrasilSuperbomemcasaCrawler.java
package br.com.lett.crawlernode.crawlers.corecontent.brasil;

import br.com.lett.crawlernode.core.models.Card;
import br.com.lett.crawlernode.core.models.Product;
import br.com.lett.crawlernode.core.models.ProductBuilder;
import br.com.lett.crawlernode.core.session.Session;
import br.com.lett.crawlernode.core.task.impl.Crawler;
import br.com.lett.crawlernode.util.CrawlerUtils;
import br.com.lett.crawlernode.util.Logging;
import com.google.common.collect.Sets;
import exceptions.MalformedPricingException;
import exceptions.OfferException;
import models.Offer;
import models.Offers;
import models.pricing.*;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

public class BrasilSuperbomemcasaCrawler extends Crawler {

   private static final String HOME_PAGE = "https://www.redesuperbom.com.br";
   protected Set<String> cards = Sets.newHashSet(Card.ELO.toString(), Card.VISA.toString(), Card.MASTERCARD.toString(), Card.AMEX.toString(), Card.HIPERCARD.toString());

   private static final String SELLER_NAME = "Superbom";

   public BrasilSuperbomemcasaCrawler(Session session) {
      super(session);
      }

   @Override
   public boolean shouldVisit() {
      String href = this.session.getOriginalURL().toLowerCase();
      return !FILTERS.matcher(href).matches() && (href.startsWith(HOME_PAGE));
   }

   @Override
   public List<Product> extractInformation(Document doc) throws Exception {
      super.extractInformation(doc);
      List<Product> products = new ArrayList<>();

      Element productElement = doc.selectFirst(".product-detail");

      if (isProductPage(productElement)) {
         Logging.printLogDebug(logger, session, "Product page identified: " + this.session.getOriginalURL());

         String internalId = CrawlerUtils.scrapStringSimpleInfoByAttribute(productElement, ".page-product__container", "data-id");
         String internalPid = internalId;
         String name = CrawlerUtils.scrapStringSimpleInfo(productElement, ".product-details__name", false);
         String primaryImage = scrapImage(productElement);
         Offers offers = scrapOffers(productElement);

         // Creating the product
         Product product = ProductBuilder.create()
            .setUrl(session.getOriginalURL())
            .setInternalId(internalId)
            .setInternalPid(internalPid)
            .setName(name)
            .setPrimaryImage(primaryImage)
            .setOffers(offers)
            .build();

         products.add(product);
      } else {
         Logging.printLogDebug(logger, session, "Not a product page " + this.session.getOriginalURL());
      }

      return products;
   }

   private String scrapImage(Element productElement) {
      String imgUrl = CrawlerUtils.scrapStringSimpleInfoByAttribute(productElement, "img.js-product-galery__photo", "src");

      return imgUrl != null ? imgUrl.replace("media", "uploads").replaceFirst("/\\d{3,4}x\\d{3,4}", "") : null;
   }

   private Offers scrapOffers(Element productElement) throws OfferException, MalformedPricingException {
      Offers offers = new Offers();
      Pricing pricing = scrapPricing(productElement);
      List<String> sales = scrapSales(pricing);


      offers.add(Offer.OfferBuilder.create()
         .setUseSlugNameAsInternalSellerId(true)
         .setSellerFullName(SELLER_NAME)
         .setIsBuybox(false)
         .setIsMainRetailer(true)
         .setPricing(pricing)
         .setSales(sales)
         .build());

      return offers;
   }

   private List<String> scrapSales(Pricing pricing) {
      List<String> sales = new ArrayList<>();
      String salesOnJson = CrawlerUtils.calculateSales(pricing);
      sales.add(salesOnJson);
      return sales;
   }

   private Pricing scrapPricing(Element productElement) throws MalformedPricingException {

      Double spotlightPrice = CrawlerUtils.scrapDoublePriceFromHtml(productElement, ".product-details__price-por", null, false, ',', session);
      Double priceFrom = CrawlerUtils.scrapDoublePriceFromHtml(productElement, ".product-details__price-de", null, false, ',', session);;

      CreditCards creditCards = scrapCreditCards(spotlightPrice);
      BankSlip bankSlip = BankSlip.BankSlipBuilder.create()
         .setFinalPrice(spotlightPrice)
         .build();

      return Pricing.PricingBuilder.create()
         .setSpotlightPrice(spotlightPrice)
         .setPriceFrom(priceFrom)
         .setCreditCards(creditCards)
         .setBankSlip(bankSlip)
         .build();
      }

   private CreditCards scrapCreditCards(Double spotlightPrice) throws MalformedPricingException {
      CreditCards creditCards = new CreditCards();

      Installments installments = new Installments();
      installments.add(Installment.InstallmentBuilder.create()
         .setInstallmentNumber(1)
         .setInstallmentPrice(spotlightPrice)
         .build());

      for (String card : cards) {
         creditCards.add(CreditCard.CreditCardBuilder.create()
            .setBrand(card)
            .setInstallments(installments)
            .setIsShopCard(false)
            .build());
      }

      return creditCards;
   }

   private boolean isProductPage(Element product) {
      return product != null;
   }
}

File: src/java/br/com/lett/crawlernode/crawlers/corecontent/brasil/BrasilSuperbomemcasaCrawler.kt
package br.com.lett.crawlernode.crawlers.corecontent.brasil

import br.com.lett.crawlernode.core.fetcher.models.Request
import br.com.lett.crawlernode.core.models.Card
import br.com.lett.crawlernode.core.models.Product
import br.com.lett.crawlernode.core.models.ProductBuilder
import br.com.lett.crawlernode.core.session.Session
import br.com.lett.crawlernode.core.task.impl.Crawler
import br.com.lett.crawlernode.crawlers.corecontent.belgium.BelgiumDelhaizeCrawler
import br.com.lett.crawlernode.util.*
import models.Offer
import models.Offers
import models.pricing.Pricing
import org.json.JSONArray
import org.json.JSONObject

/**
 * Date: 21/12/20
 *
 * @author Fellype Layunne
 *
 */
class BrasilSuperbomemcasaCrawler (session: Session) : Crawler(session){

   companion object {
      const val SELLER_NAME: String = "Superbomemcasa"

      fun scrapAttributes(jsonArray: JSONArray, key: String): String? {
         for (att in jsonArray) {
            if (att is JSONObject) {
               if (att.optString("attribute_code") == key) {
                  return att.optString("value")
               }
            }
         }
         return null
      }
   }

   private fun scrapInternalPidFromUrl(): String {

      return session.originalURL.substringAfterLast(".").substringBeforeLast("p")
   }

   override fun fetch(): JSONObject {

      val url =   "https://sb.superbomemcasa.com.br/rest/default/V1/products/" + scrapInternalPidFromUrl()

      val headers = HashMap<String, String>()

      val request = Request.RequestBuilder.create()
         .setUrl(url)
         .setHeaders(headers)
         .build()

      return dataFetcher.get(session, request).body.toJson()
   }

   override fun extractInformation(json: JSONObject): MutableList<Product> {

      if (!isProductPage()) {
         return mutableListOf()
      }

      val name = json.optString("name")
      val internalId = json.optString("id")
      val internalPid = json.optString("sku")

      val jsonAtt = JSONUtils.getJSONArrayValue(json, "custom_attributes")

      val description = scrapAttributes(jsonAtt, "description")

      var primaryImage = scrapAttributes(jsonAtt, "swatch_image")

      if (primaryImage != null) {
         primaryImage = "https://superbom.s3-sa-east-1.amazonaws.com/catalog/product${primaryImage}"
      }

      val offers = scrapOffers(json, jsonAtt)

      val product = ProductBuilder()
         .setUrl(session.originalURL)
         .setInternalId(internalId)
         .setInternalPid(internalPid)
         .setName(name)
         .setPrimaryImage(primaryImage)
         .setDescription(description)
         .setOffers(offers)
         .build()

      return mutableListOf(product)
   }

   private fun scrapOffers(json: JSONObject, jsonAtt: JSONArray): Offers {

      val offers = Offers()

      if (json.optInt("status") != 1) {
         return offers
      }

      val price = json.optDouble("price")

      val specialPrice = scrapAttributes(jsonAtt, "special_price")?.toDouble()

      val spotlightPrice = specialPrice ?: price

      val priceFrom = if (specialPrice == null) null else price

      val bankSlip = spotlightPrice.toBankSlip()

      val creditCards = listOf(
         Card.MASTERCARD,
         Card.VISA,
         Card.HIPERCARD,
         Card.ELO,
         Card.DINERS,
         Card.AMEX,
         Card.SOROCRED,
         Card.CABAL,
         Card.JCB
      ).toCreditCards(price)

      offers.add(
         Offer.OfferBuilder.create()
            .setPricing(
               Pricing.PricingBuilder.create()
                  .setSpotlightPrice(spotlightPrice)
                  .setPriceFrom(priceFrom)
                  .setCreditCards(creditCards)
                  .setBankSlip(bankSlip)
                  .build()
            )
            .setIsMainRetailer(true)
            .setIsBuybox(false)
            .setUseSlugNameAsInternalSellerId(true)
            .setSellerFullName(SELLER_NAME)
            .setSales(listOf())
            .build()
      )

      return offers
   }

   private fun isProductPage(): Boolean {
      return scrapInternalPidFromUrl().isNotEmpty()
   }

}
